modules/apps/29-fee/module_test.go

```diff
@@ -0,0 +1,245 @@
+package fee_test
+
+import (
+	"github.com/cosmos/ibc-go/modules/apps/29-fee/types"
+	transfertypes "github.com/cosmos/ibc-go/modules/apps/transfer/types"
+	channeltypes "github.com/cosmos/ibc-go/modules/core/04-channel/types"
+	host "github.com/cosmos/ibc-go/modules/core/24-host"
+)
+
+func (suite *FeeTestSuite) TestOnChanOpenInit() {
+	testCases := []struct {
+		name    string
+		version string
+		expPass bool
+	}{
+		{
+			"valid fee middleware and transfer version",
+			"fee29-1:ics20-1",
+			true,
+		},
+		{
+			"invalid fee middleware version",
+			"otherfee28-1:ics20-1",
+			false,
+		},
+		{
+			"invalid transfer version",
+			"fee29-1:wrongics20-1",
+			false,
+		},
+		{
+			"incorrect wrapping delimiter",
+			"fee29-1//ics20-1",
+			false,
+		},
+		{
+			"transfer version not wrapped",
+			"fee29-1",
+			false,
+		},
+		{
+			"fee version not included",
+			"ics20-1",
+			false,
+		},
+		{
+			"hanging delimiter",
+			"fee29-1:ics20-1:",
+			false,
+		},
+	}
+
+	for _, tc := range testCases {
+		ctx := suite.chainA.GetContext()
+		cap, err := suite.chainA.GetSimApp().ScopedIBCKeeper.NewCapability(ctx, tc.name)
+		suite.Require().NoError(err)
+		err = suite.module.OnChanOpenInit(
+			ctx,
+			channeltypes.UNORDERED,
+			[]string{"connection-1"},
+			transfertypes.FeePortID,
+			"channel-1",
+			cap,
+			channeltypes.NewCounterparty(transfertypes.FeePortID, "channel-2"),
+			tc.version,
+		)
+
+		if tc.expPass {
+			suite.Require().NoError(err, "unexpected error from version: %s", tc.version)
+
+			// check that capabilities are properly claimed and issued
+			ctx := suite.chainA.GetContext()
+			ibcCap, ok := suite.chainA.GetSimApp().ScopedIBCFeeKeeper.GetCapability(ctx, host.ChannelCapabilityPath(transfertypes.FeePortID, "channel-1"))
+			suite.Require().NotNil(ibcCap, "IBC capability is nil on fee keeper")
+			suite.Require().True(ok)
+
+			appFeeCap, ok := suite.chainA.GetSimApp().ScopedIBCFeeKeeper.GetCapability(ctx, types.AppCapabilityName(transfertypes.FeePortID, "channel-1"))
+			suite.Require().NotNil(appFeeCap, "App capability not created or owned by ibc fee keeper")
+			suite.Require().True(ok)
+			transferCap, ok := suite.chainA.GetSimApp().ScopedTransferKeeper.GetCapability(ctx, host.ChannelCapabilityPath(transfertypes.FeePortID, "channel-1"))
+			suite.Require().NotNil(transferCap, "App capability not claimed by transfer keeper")
+			suite.Require().True(ok)
+			suite.Require().Equal(appFeeCap, transferCap, "app capabilities not equal")
+		} else {
+			suite.Require().Error(err, "error not returned for version: %s", tc.version)
+		}
+	}
+}
+
+func (suite *FeeTestSuite) TestOnChanOpenTry() {
+	testCases := []struct {
+		name      string
+		version   string
+		cpVersion string
+		capExists bool
+		expPass   bool
+	}{
+		{
+			"valid fee middleware and transfer version",
```

Comment:
can you add a test where the fee and transfer versions are valid but don't match (maybe have an extra nested version)